ngraph/core/src/op/interpolate.cpp

```diff
@@ -111,143 +117,518 @@ namespace ngraph
+            scales.insert(scales.end(), scales_ptr, scales_ptr + num_of_axes);
+        }
+        else
+        {
+            auto target_shape = get_target_shape_vector(args, num_of_axes);
+            for (std::size_t i = 0; i < num_of_axes; ++i)
+            {
+                std::size_t axis = axes[i];
+                float scale =
+                    static_cast<float>(target_shape[i]) / static_cast<float>(input_shape[axis]);
+                scales.push_back(scale);
+            }
+        }
+        return scales;
+    }
+
+    size_t_vector get_padded_input_shape(const Shape& input_shape,
+                                         const size_t_vector& pads_begin,
+                                         const size_t_vector& pads_end)
+    {
+        size_t_vector result = input_shape;
+        std::size_t rank = input_shape.size();
+        for (std::size_t i = 0; i < rank; ++i)
+        {
+            result[i] += pads_begin[i] + pads_end[i];
+        }
+
+        return result;
+    }
+
+    size_t_vector shape_infer_with_scales(const size_t_vector& padded_shape,
+                                          const int64_vector& axes,
+                                          const std::vector<float>& scales)
+    {
+        auto out_shape = padded_shape;
+        std::size_t num_of_axes = axes.size();
+        for (std::size_t i = 0; i < num_of_axes; ++i)
+        {
+            int64_t axis = axes[i];
+            float scaled_len = static_cast<float>(padded_shape[axis]) * scales[i];
+            out_shape[axis] = static_cast<std::size_t>(scaled_len);
+        }
+        return out_shape;
+    }
+
+    size_t_vector shape_infer_with_target_shape(const size_t_vector& padded_shape,
+                                                const int64_vector& axes,
+                                                const int64_vector& target_shape)
+    {
+        auto out_shape = padded_shape;
+        std::size_t num_of_axes = axes.size();
+        for (std::size_t i = 0; i < num_of_axes; ++i)
+        {
+            out_shape[axes[i]] = target_shape[i];
+        }
+        return out_shape;
+    }
+
+    template <typename T>
+    std::vector<T> correct_pad(const std::vector<T>& p, std::size_t rank)
+    {
+        std::size_t pad_len = p.size();
+        if (pad_len == rank)
+        {
+            return p;
+        }
+
+        std::vector<T> result;
+
+        if (pad_len > rank)
+        {
+            result.insert(result.end(), p.begin(), p.begin() + rank);
+        }
+        else
+        {
+            result = p;
+            result.insert(result.end(), rank - pad_len, T{});
+        }
+
+        return result;
+    }
+
+    template <element::Type_t ET>
+    inline bool evaluate(const HostTensorVector& args,
+                         const HostTensorPtr& out,
+                         const op::v4::Interpolate::InterpolateAttrs& attrs)
+    {
+        using T = typename element_type_traits<ET>::value_type;
```

Comment:
Looks like we can avoid template arguments for this code.

Reply:
No. This function calls a template function using `T` as a template parameter.

Reply:
At the current approach each instantiation of this code impacts the binary size.

But you need the template argument only in the last line (I think you can use CoordinateTranspose instead of NDimArrayView<T>).

Moreover the T or ET types you can get from the host tensor.

Reply:
Reduced size of this template function.